velox/experimental/cudf/exec/ExpressionEvaluator.cpp

```diff
@@ -713,6 +780,119 @@ class CardinalityFunction : public CudfFunction {
+            expr->inputs()[1])) {
+      auto constValue = constExpr->value();
+      right_ = VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH(
+          createCudfScalar, constValue->typeKind(), constValue);
+    }
+
+    VELOX_CHECK(
+        !(left_ != nullptr && right_ != nullptr),
+        "Not support both left and right are literals");
+  }
+
+  ColumnOrView eval(
+      std::vector<ColumnOrView>& inputColumns,
+      rmm::cuda_stream_view stream,
+      rmm::device_async_resource_ref mr) const override {
+    auto inputCol = asView(inputColumns[0]);
+    if (left_ == nullptr && right_ == nullptr) {
+      return cudf::binary_operation(
+          asView(inputColumns[0]),
+          asView(inputColumns[1]),
+          op_,
+          type_,
+          stream,
+          mr);
+    } else if (left_ == nullptr) {
+      return cudf::binary_operation(
+          asView(inputColumns[0]), *right_, op_, type_, stream, mr);
+    }
+    return cudf::binary_operation(
+        *left_, asView(inputColumns[0]), op_, type_, stream, mr);
+  }
+
+ private:
+  const cudf::binary_operator op_;
+  const cudf::data_type type_;
+  std::unique_ptr<cudf::scalar> left_;
+  std::unique_ptr<cudf::scalar> right_;
+};
+
+class SwitchFunction : public CudfFunction {
+ public:
+  SwitchFunction(const std::shared_ptr<velox::exec::Expr>& expr)
+      : constantScalars_(collectScalars(expr)) {
+    VELOX_CHECK_EQ(
+        expr->inputs().size(), 3, "case when expects exactly 3 inputs");
+    VELOX_CHECK_EQ(
+        expr->inputs()[0]->type()->kind(),
+        TypeKind::BOOLEAN,
+        "The switch condition result type should be boolean");
+    bool hasNonLiteral = false;
+    for (const auto& expr : expr->inputs()) {
+      if (std::dynamic_pointer_cast<velox::exec::ConstantExpr>(expr)) {
+        continue;
+      }
+      hasNonLiteral = true;
+      break;
+    }
+    VELOX_CHECK(
+        hasNonLiteral,
+        "The switch expression should have at least 1 non-literal column");
+  }
+
+  // Maybe this is a common case, all the column requires to be column, though
+  // it is literal. To refactor then if this is a common case.
+  ColumnOrView eval(
+      std::vector<ColumnOrView>& inputColumns,
+      rmm::cuda_stream_view stream,
+      rmm::device_async_resource_ref mr) const override {
+    const auto numRows = asView(inputColumns[0]).size();
+    std::vector<cudf::column_view> resolvedInputs;
+    std::vector<std::unique_ptr<cudf::column>>
+        constantColumns; // Reserve for memory.
+    for (auto i = 0, j = 0; i < constantScalars_.size(); ++i) {
+      if (constantScalars_[i] != nullptr) {
+        auto& scalar = *constantScalars_[i];
+        auto column =
+            cudf::make_column_from_scalar(scalar, numRows, stream, mr);
+        resolvedInputs.push_back(column->view());
+        constantColumns.push_back(std::move(column));
+      } else {
+        resolvedInputs.push_back(asView(inputColumns[j++]));
+      }
```

Comment:
This may not be necessary as I can see cudf supports constants in `copy_if_else` via scalars. [source](https://github.com/rapidsai/cudf/blob/706164d28c73d99a4db4da475677e869e525ff6b/cpp/src/copying/copy.cu#L429-L449)